doc/source/_ext/suppress.py

```diff
@@ -0,0 +1,291 @@
+        suppress_records = [('sphinx.ext.intersphinx', '.*')]
+"""
+
+__all__ = ()
+
+import abc
+import importlib
+import inspect
+import itertools
+import logging
+import pkgutil
+import re
+import warnings
+from typing import TYPE_CHECKING
+
+from sphinx.util.logging import NAMESPACE, SphinxLoggerAdapter
+
+if TYPE_CHECKING:
+    from sphinx.application import Sphinx
+    from sphinx.config import Config
+    from sphinx.extension import Extension
+
+
+def partition(predicate, iterable):
+    """
+    Return a pair `(no, yes)`, where *yes* and *no* are subsets of *iterable*
+    over which *predicate* evaluates to |False| and |True| respectively.
+    """
+
+    no, yes = itertools.tee(iterable)
+    no, yes = itertools.filterfalse(predicate, no), filter(predicate, yes)
+    return no, yes
+
+
+def notnone(value):
+    return value is not None
+
+
+ALL = object()
+
+
+def _normalize_level(level):
+    if isinstance(level, int):
+        return level
+
+    # pylint: disable-next=W0212
+    result = logging._nameToLevel.get(level)
+    if result is not None:
+        return result
+
+    return None  # unknown level
+
+
+def _parse_levels(levels):
+    if not isinstance(levels, (list, tuple)):
+        if isinstance(levels, (int, str)):
+            levels = [levels]
+    return list(filter(notnone, map(_normalize_level, levels)))
+
+
+class SphinxSuppressFilter(logging.Filter, metaclass=abc.ABCMeta):
+    def filter(self, record):
+        # type: (logging.LogRecord) -> bool
+        return not self.suppressed(record)
+
+    @abc.abstractmethod
+    def suppressed(self, record):
+        # type: (logging.LogRecord) -> bool
+        pass
+
+
+class SphinxSuppressLogger(SphinxSuppressFilter):
+    r"""A filter suppressing logging records issued by a Sphinx logger."""
+
+    def __init__(self, name, levels=()):
+        """
+        Construct a :class:`SphinxSuppressLogger`.
+
+        :param name: The (real) logger name to suppress.
+        :type name: str
+        :param levels: Optional logging levels to suppress.
+        :type levels: bool | list[int]
+        """
+
+        super().__init__(name)
+        if isinstance(levels, bool):
+            self.levels = ALL if levels else []
+        else:
+            self.levels = _parse_levels(levels)
+
+    def suppressed(self, record):
+        return (
+            logging.Filter.filter(self, record)
+            and (self.levels is ALL or record.levelno in self.levels)
+        )
+
+
+class SphinxSuppressPatterns(SphinxSuppressFilter):
+    r"""A filter suppressing matching messages."""
+
+    def __init__(self, patterns=()):
+        """
+        Construct a :class:`SphinxSuppressPatterns`.
+
+        :param patterns: Optional logging messages (regex) to suppress.
+        :type patterns: list[str | re.Pattern]
+        """
+
+        super().__init__('')  # all loggers
+        self.patterns = set(map(re.compile, patterns))
+
+    def suppressed(self, record):
```

Comment:
If you want to reduce the complexity of this call, we can check whether `self.patterns` is empty beforehand as follows:

```python
        if not self.patterns:
            return False

        m = record.getMessage()
        return any(p.search(m) for p in self.patterns)
```